aws-core/src/main/java/com/amplifyframework/auth/AWSCredentialsProvider.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.auth
+
+import aws.smithy.kotlin.runtime.auth.awscredentials.Credentials
+import aws.smithy.kotlin.runtime.auth.awscredentials.CredentialsProvider
+import com.amplifyframework.core.Consumer
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+/**
+ * Customer provided CredentialsProvider implementation that fetches and returns AWSCredentials or a subclass such
+ * as AWSTemporaryCredentials. For example:
+ *
+ * class AWSTemporaryCredentialsProvider: AWSCredentialsProvider<AWSTemporaryCredentials> {
+ *   override fun fetchAWSCredentials(
+ *     onSuccess: Consumer<AWSTemporaryCredentials>,
+ *     onError: Consumer<AuthException>
+ *   ) {
+ *     // customer provided fetch implementation
+ *   }
+ * }
+ */
+interface AWSCredentialsProvider<out T : AWSCredentials> {
+
+    fun fetchAWSCredentials(
+        onSuccess: Consumer<@UnsafeVariance T>,
+        onError: Consumer<AuthException>
+    )
+}
+
+fun <T : AWSCredentials> convertToSdkCredentialsProvider(
```

Comment:
I dont think this needs to be publicly exposed. If the goal is to pass AWSCredentialsProvider to other components and SDK CredentialsProvider is always accessible through escape hatch, I dont see a reason for this conversion API. Maybe start with non-public and if customers need it we can exposed it in the future.

Reply:
@div5yesh The reason I exposed is because I don't think we want customers having to use multiple credentials providers, or at least, having to create their own AWSCredentialsProvider and SDKs CredentialsProvider.

Our suggested implementation should be to create a single AWSCredentialsProvider, and when that is needed for the Kotlin SDK escape hatch, they can use this public conversion method.